src/box/txn_limbo.c

```diff
@@ -553,50 +751,215 @@ txn_limbo_read_rollback(struct txn_limbo *limbo, int64_t lsn)
-txn_limbo_write_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term)
+static void
+txn_limbo_make_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term,
+		       struct synchro_request *req)
 {
-	assert(latch_is_locked(&limbo->promote_latch));
 	/*
 	 * We make sure that promote is only written once everything this
 	 * instance has may be confirmed.
 	 */
 	struct txn_limbo_entry *e = txn_limbo_last_synchro_entry(limbo);
 	assert(e == NULL || e->lsn <= lsn);
 	(void) e;
-	struct synchro_request req = {
+
+	/*
+	 * The PROMOTE request we're about to write depends on
+	 * the previous one if it exists. It's an inductive property.
+	 */
+	struct rlist *queue = &limbo->pending_promotes;
+	struct pending_promote *last = !rlist_empty(queue)
+		? rlist_last_entry(queue, typeof(*last), link)
+		: NULL;
+
+	uint64_t greatest_term = limbo->promote_greatest_term;
+	*req = (struct synchro_request) {
 		.type = IPROTO_RAFT_PROMOTE,
-		.replica_id = limbo->owner_id,
 		.origin_id = instance_id,
-		.lsn = lsn,
+		.wait_ack = true,
 		.term = term,
+		/* These fields require extra tinkering with: */
+		.prev_term  = last ? last->req.term : greatest_term,
+		.replica_id = last ? last->req.origin_id : limbo->owner_id,
+		.lsn        = last ? last->req.self_lsn : lsn,
 		/*
 		 * Confirmed_vclock is only persisted in checkpoints. It doesn't
 		 * appear in WALs and replication.
 		 */
 		.confirmed_vclock = NULL,
 	};
-	if (txn_limbo_req_prepare(limbo, &req) < 0)
-		return -1;
-	synchro_request_write(&req);
+}
+
+int
+txn_limbo_write_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+	int rc = -1;
+
+	struct synchro_request req;
+	txn_limbo_make_promote(limbo, lsn, term, &req);
+
+	limbo->is_writing_promote = true;
+	if (txn_limbo_req_prepare(limbo, &req) < 0) {
+		limbo->is_writing_promote = false;
+		goto end;
+	}
+	req.self_lsn = synchro_request_write(&req);
+	ERROR_INJECT(ERRINJ_LIMBO_WRITE_PROMOTE_SLEEP, { fiber_sleep(1); });
+	say_info("PROMOTE: write %s", synchro_request_to_string(&req));
 	txn_limbo_req_commit(limbo, &req);
+	limbo->is_writing_promote = false;
+
+	/* Immediately acknowledge our own write. */
+	int64_t prev_lsn, self_lsn = req.self_lsn;
+	ERROR_INJECT_YIELD(ERRINJ_LIMBO_PROMOTE_ACK_SLEEP);
+	txn_limbo_ack_already_seen(limbo, instance_id, self_lsn, &prev_lsn);
+	txn_limbo_ack_promote(limbo, instance_id, self_lsn, prev_lsn);
+
+	rc = 0;
+end:
+	/* Notify possible waiters. */
+	limbo->write_promote_count++;
+	fiber_cond_broadcast(&limbo->write_promote_cond);
+
+	return rc;
+}
+
+uint64_t
+txn_limbo_promote_attempts(struct txn_limbo *limbo)
+{
+	return limbo->write_promote_count;
+}
+
+int
+txn_limbo_wait_promote_attempts(struct txn_limbo *limbo,
+				uint64_t count, double timeout)
+{
+	struct fiber_cond *cond = &limbo->write_promote_cond;
+	while (limbo->write_promote_count < count)
+		if (fiber_cond_wait_timeout(cond, timeout) != 0)
+			return -1;
 	return 0;
```

Comment:
This won't hold if the calling fiber would be simply woken up manually via `fiber.wakeup()`. Why do we need to wait for it this way? I can only imagine places which don't need to wait at all or need to wait for a timeout, not for a count.